drivers/adc/adc_gecko.c

```diff
@@ -0,0 +1,376 @@
+		return -EINVAL;
+	}
+
+	if (channel_cfg->differential) {
+		LOG_ERR("Differential channels are not (yet) supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Invalid channel gain");
+		return -EINVAL;
+	}
+
+	/* Setup input */
+	config_ch = CH_CFG(channel_cfg->channel_id);
+	config_ch->input_positive = channel_cfg->input_positive;
+
+	/* Setup acquistion time */
+	acq_time_index = adc_gecko_check_acq_time(
+		channel_cfg->acquisition_time);
+	if (acq_time_index < 0) {
+		return acq_time_index;
+	}
+	config_ch->acq_time = acq_time_index;
+
+	/* Setup reference */
+	switch (channel_cfg->reference) {
+	case ADC_REF_VDD_1:
+	case ADC_REF_INTERNAL:
+		config_ch->reference = adcRef2V5;
+		break;
+
+	case ADC_REF_VDD_1_2:
+		config_ch->reference = adcRef1V25;
+		break;
+
+	default:
+		LOG_ERR("Invalid channel reference");
+		return -EINVAL;
+	}
+
+	LOG_DBG("Channel setup succeeded!");
+
+	return 0;
+}
+
+static int adc_gecko_init(struct device *dev)
+{
+	const struct adc_gecko_cfg *config = DEV_CFG(dev);
+	struct adc_gecko_data *data = DEV_DATA(dev);
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+
+	LOG_DBG("Initializing....");
+
+	data->dev = dev;
+
+	/* Enable ADC clock */
+	CMU_ClockEnable(config->clock, true);
+
+	/* Base the ADC configuration on the default setup. */
+	ADC_Init_TypeDef init = ADC_INIT_DEFAULT;
+
+	/* Initialize timebases */
+	init.timebase = ADC_TimebaseCalc(0);
+	init.prescale = (config->prescaler - 1);
+	ADC_Init(adc, &init);
+
+	config->irq_cfg_func();
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+static const struct adc_driver_api api_gecko_driver_api = {
+	.channel_setup = adc_gecko_channel_setup,
+	.read = adc_gecko_read,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = adc_gecko_read_async,
+#endif
+	.ref_internal = 2500,
+};
+
+#define GECKO_ADC_INIT(index)						\
+									\
+	static void adc_gecko_cfg_func_##index(void);			\
+									\
+	static const struct adc_gecko_cfg adc_gecko_cfg_##index = {	\
+		.base = (ADC_TypeDef *)					\
+			DT_INST_##index##_SILABS_GECKO_ADC_BASE_ADDRESS, \
+		.irq_cfg_func = adc_gecko_cfg_func_##index,		\
+		.clock = CLOCK_ADC(					\
+			DT_INST_##index##_SILABS_GECKO_ADC_PERIPHERAL_ID), \
+		.prescaler = DT_INST_##index##_SILABS_GECKO_ADC_PRESCALER \
+	};								\
+	static struct adc_gecko_data adc_gecko_data_##index = {		\
+		ADC_CONTEXT_INIT_TIMER(adc_gecko_data_##index, ctx),	\
+		ADC_CONTEXT_INIT_LOCK(adc_gecko_data_##index, ctx),	\
+		ADC_CONTEXT_INIT_SYNC(adc_gecko_data_##index, ctx),	\
+	};								\
+									\
+	DEVICE_AND_API_INIT(adc_##index,				\
+				DT_INST_##index##_SILABS_GECKO_ADC_LABEL, \
+				&adc_gecko_init, &adc_gecko_data_##index, \
+				&adc_gecko_cfg_##index, POST_KERNEL,	\
+				CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,	\
+				&api_gecko_driver_api);			\
+									\
+	static void adc_gecko_cfg_func_##index(void)			\
+	{								\
+		IRQ_CONNECT(DT_INST_##index##_SILABS_GECKO_ADC_IRQ_##index, \
```

Comment:
It should be `_IRQ_0`, not `_IRQ_##index`, as this is not the index of the ADC instance, but the index of the interrupt defined for this instance (and Gecko ADCs use only one interrupt, right?).
Same applies to `_IRQ_##index##_PRIORITY` below.

Reply:
Nice catch, I may have been a bit too copy->paste happy here. And you are correct, each ADC peripheral has only 1 interrupt assigned to it.
Even though the `##index` might seem a bit overdoing it here as the Pearl/Jade Geckos only feature one ADC peripheral, I do plan on adding board support for the Giant Gecko series 1 as well, which actually features 2 ADC peripherals. Hence, why I already used a generic macro for initializing the ADC context.